src/USER-HSMA/hsma3d.cpp

```diff
@@ -0,0 +1,1945 @@
+
+	  AdjustParticle_Double(X, nlocal, Lx, Ly, Lz);
+
+	  MPI_Iallgatherv((double*)X, nlocal * 3, MPI_DOUBLE, (double*)AllSource, nlocal_All, Size_All, MPI_DOUBLE, world, &request);
+	  MPI_Iallgatherv((double*)Q, nlocal, MPI_DOUBLE, (double*)AllQ, nlocal_All_Q, Size_All_Q, MPI_DOUBLE, world, &request_Q);
+
+	  //Find the image charge
+	  int lx = ceil((Rs - Lx / 2) / Lx), ly = ceil((Rs - Ly / 2) / Ly), lz = ceil((Rs - Lz / 2) / Lz);
+	  int TotalNumber = ceil(nlocal * (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1) / 2);
+	  double ImageCharge[TotalNumber][4];
+	  int ImageNumber;
+	  SetImageCharge(ImageCharge, &ImageNumber, (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1), X, Q, nlocal, Rs, Lx, Ly, Lz, lx, ly, lz);
+
+	  //Calculate near field potential which contains all the contributions in the cut-off sphere B at the monitoring points
+	  double Near[Nw], Near_All[Nw];
+
+	  //CalculateNearFieldAndZD(Near, ImageCharge, ImageNumber, Nw, IF_FMM_RightTerm, Lx, Ly, Lz, PointSum_New, QuizSum_New);
+	  CalculateNearFieldAndZD(Near, ImageCharge, ImageNumber, Nw, IF_FMM_RightTerm, Lx, Ly, Lz, PointSum, QuizSum,tolerance);
+	  MPI_Reduce(Near, Near_All, Nw, MPI_DOUBLE, MPI_SUM, 0, world);
+
+	  //Solve The Least Square Problem
+	  double C[p * p];
+	  if (me == 0) { SolveLeastSquareProblem(C, A, Near_All, p, Nw); }
+	  MPI_Bcast(C, p * p, MPI_DOUBLE, 0, world);
+	  MPI_Wait(&request, &status);
+	  MPI_Wait(&request_Q, &status);
+
+	  //Compute final force, potential and energy
+	  double Energy_HSMA;
+	  double Force[nlocal][3], Pot[nlocal];
+	  TotalNumber = ceil(maxatom * (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1) / 2);//¼õÉÙ´æ´¢Á¿ÏûºÄ
+	  double ImageCharge_All[TotalNumber][4];
+	  SetImageCharge(ImageCharge_All, &ImageNumber, (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1), AllSource, AllQ, maxatom, Rs, Lx, Ly, Lz, lx, ly, lz);
+
+	  Energy_HSMA = FinalCalculateEnergyAndForce(Force, Pot, X, Q, nlocal, ImageCharge_All, ImageNumber, Fibonacci, (double**)QRD, (double**)QLocalRD, C, p, Fp, F, Rs, pi, IF_FMM_FinalPotential,tolerance);
+
+	  scale = 1.0;
+	  const double qscale = qqrd2e * scale;
+
+	  //Assemble results
+	  for (int i = 0; i < nlocal; i++)
+	  {
+		  f[i][0] += Force[i][0] * qscale;
+		  f[i][1] += Force[i][1] * qscale;
+		  f[i][2] += Force[i][2] * qscale;
+	  }
+
+	  double Energy_HSMA_All;
+	  MPI_Allreduce(&Energy_HSMA, &Energy_HSMA_All, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+
+	  if (eflag_global) {
+		  energy += Energy_HSMA_All * qscale / 2;
+	  }
+
+	  if (evflag_atom) {
+		  if (eflag_atom) {
+			  for (int i = 0; i < nlocal; i++) {
+				  eatom[i] += Pot[i];
+				  eatom[i] *= qscale;
+			  }
+		  }
+		  if (vflag_atom)
+			  for (int i = 0; i < nlocal; i++)
+				  for (int j = 0; j < 6; j++) vatom[i][j] *= q[i] * qscale;
+	  }
+  }
+}
+
+double HSMA3D::memory_usage()
+{
+  double bytes = 0.0;
+  bytes += maxatom * sizeof(double);
+  bytes += 3*maxatom * sizeof(double);
+  return bytes;
+}
+
+double fac(double t)//calculate factorial
```

Comment:
An optimized factorial function is available in the `MathSpecial` namespace.

Reply:
Include “math_special.h” and then return factorial()